lib/Dataflow_analysis.ml

```diff
@@ -0,0 +1,800 @@
+open Core_kernel
+open Mir
+
+(***********************************)
+(* Basic datatypes                 *)
+(***********************************)
+
+(**
+   A label is a unique identifier for a node in the dataflow/dependency graph, and
+   usually corresponds to one node in the Mir.
+*)
+type label = int [@@deriving sexp]
+
+(**
+   A 'reaching dependency' (or reaching_dep or RD) statement (v, l) says that the variable
+   v could have been affected at the label l.
+*)
+type reaching_dep = expr * int [@@deriving sexp, hash, compare]
+
+(**
+   A reaching dependency set holds the set of reaching dependencies that could be true at
+   some point.
+*)
+module ReachingDepSet = Set.Make (struct
+  type t = reaching_dep
+
+  let compare : reaching_dep -> reaching_dep -> int = compare
+  let sexp_of_t = sexp_of_reaching_dep
+  let t_of_sexp = reaching_dep_of_sexp
+end)
+
+module LabelMap = Map.Make (struct
+  type t = label
+
+  let compare : int -> int -> int = compare
+  let sexp_of_t = sexp_of_int
+  let t_of_sexp = int_of_sexp
+end)
+
+module LabelSet = Set.Make (struct
+  type t = label
+
+  let compare : int -> int -> int = compare
+  let sexp_of_t = sexp_of_int
+  let t_of_sexp = int_of_sexp
+end)
+
+module ExprSet = Set.Make (struct
+  type t = expr
+
+  let compare : expr -> expr -> int = compare
+  let sexp_of_t = sexp_of_expr
+  let t_of_sexp = expr_of_sexp
+end)
+
+(**
+   Description of where a node in the dependency graph came from, where MirNode is the
+   location from an Mir.loc_stmt
+ *)
```

Comment:
What's this data structure used for? And why is it specific to `expr`s?

Reply:
Refering to ExprSet? I just need a set of expressions data structure throughout the code, and this is my understanding of how to do that in OCaml. If there were an easier was that would be ideal

Reply:
Oops, I was referring to source_loc - why do we need to keep track of that in this module? And is TargetTerm referring to the same lpdf target, or is it another usage? I wasn't sure what the expr term was, hence why it's specific to `expr` vs. statements or whatever.

Reply:
Ah, my mistake. I kept the source_loc from the MIR because I thought it might be useful down the line. For example, if we wanted to have some kind of code visualization showing dependencies, we'd want to be able to translate nodes back into statements.

When a `target += term` statement is encountered, it generates one node for the whole statement, plus one node for each term that was added (it splits the RHS by + operators). This is because target statements have two kinds of effects, they change the value of target (which can be used as a variable) and the terms introduce probabilistic dependencies. TargetTerm is only used when the node represents a term, and it contains the term expression and the label of the whole-statement node.

Reply:
What does "term" mean here, and what are other things in the same category where you wouldn't use `TargetTerm`? I like the idea of breaking out the probabilistic dependencies, you'll have to show me how that works Thursday. I think I get what you're saying... to paraphrase: one of the static analysis passes in this file breaks up `target += X` MIR nodes into separate static analysis nodes, and so if you're keeping track of the location, those new break-out nodes will refer to an original TargetTerm. That said, this will probably not be the only kind of node that gets broken up into multiple nodes for the sake of static analysis (and many nodes may also get combined).

That might help motivate a conversation around the best way to reconstruct the MIR or do other things like code visualization - seems like we might just want to keep pointers to the original MIR nodes w/ metadata?

Reply:
Term as in a summand of the target, which translate to independent factors of the joint. I wouldn't use TargetTerm for a node that corresponded to e.g., a variable assignment or a while loop.

Yeah, the probabilistic dependencies can be separated easily, I actually kept it as a flag in the dependency analysis functions. If you turn off all the probabilistic dependencies, you just get programmatic dependencies as they would be over the course of a single HMC step (in the case of the model block, as I understand it). 

Your paraphrase sounds right to me. So far, this is the only kind of node that gets broken up, but yeah, maybe more down the line.

Keeping pointers to the original MIR sounds good, if there were a convenient way of indexing them.